### PR TITLE
Screenings startpage

### LINKED ISSUE #2: 1 - Startsidan ska visa en lista på visningar de kommande dagarna.
**Startsidan ska visa en lista på visningar de kommande dagarna. Listan ska laddas in med hjälp av webbläsarens
fetch()
EFTER att sidan har visats, d.v.s. inte renderas på servern.**

- Endast visningar för de kommande fem dagarna ska visas

- Max 10 filmvisningar ska visas. Om det finns fler än 10 visningar de kommande fem dagarna ska bara så många dagar visas som resulterar i max 10 visningar.

### PR DESCRIPTION
**Pull request by mkor92**
## Route added: GET /api/screenings/startpage
Returns screenings with movie information.

## Serverside logic 
Filtrates so outcome is screenings for coming 5 days and max 10 screenings.

## Client-side logic
Added file movies.js - Rendering list of movies on startpage and groups
screenings for same day in own UL list.

## Styling added! For now i removed accordion on startpage, not necessary to have it for this assignment. But if we have time to fix, we can add logic to show screenings in accordion to.

Skärmbild_20230203_111756

### CODE REVIEW

**Review by jakobpalmborg — CHANGES_REQUESTED** [review]
Looks good! But as we talked about at discord, there is some problem regarding dates and language settings. The screening don't show on startpage when I have English as language in windows, but when I switch to swedish it works fine!

**Review by matodorovic — APPROVED** [review]
It works for me. It shows screening times on start page.  U have just some conflicts to be solved before merging.

**Review by jakobpalmborg — APPROVED** [review]
Great! Now it works with windows with english settings.

**Comment by mkor92** [comment]
> It works for me. It shows screening times on start page. U have just some conflicts to be solved before merging.
There was some issues with date format if computer system was in other language than swedish. It now works


> Great! Now it works with windows with english settings.
Perfect! I merge this branch now.

**Comment by mkor92** [comment]
> It works for me. It shows screening times on start page. U have just some conflicts to be solved before merging.


There was some issues with date format if computer system was in other language than swedish. It now works

**Comment by rfabricioflores** [comment]
> > It works for me. It shows screening times on start page. U have just some conflicts to be solved before merging.
> 
> There was some issues with date format if computer system was in other language than swedish. It now works

Well done!

File: src/controllers/api.controller.js [hunk 1]
```diff
@@ -1,27 +1,31 @@
...
 apiCtrl.movies = async (req, res) => {
-    res.json(await loadMovies());
-}
+  res.json(await loadMovies());
+};
 
-apiCtrl.movieScreenings = async (req, res) => { 
-    res.json(await movieScreenings(req.params.id));
-}
+apiCtrl.startpageScreenings = async (req, res) => {
```
Comment (richardolsson) [inline src/controllers/api.controller.js:12]: I see no great reason for having these "controllers". I would have just put this code in the route handlers in `api.routes.js`.

File: src/helpers/screenings.js [hunk 2]
```diff
@@ -3,13 +3,52 @@ import fetch from "node-fetch";
...
-    const res = await fetch(API);
-    const payload = await res.json();
-    return payload.data;
+  const res = await fetch(API);
+  const payload = await res.json();
+  return payload.data;
+}
+
+export async function screeningsStartpage() {
+  const apiRes = await fetch(API + "?populate=movie");
```
Comment (richardolsson) [inline src/helpers/screenings.js:18]: This will only return the first 25 screenings in the database, which are likely very old. Because all of the screenings returned are old, they will be discarded by the logic below, and no screenings show up on the home page.

File: src/helpers/screenings.js [hunk 3]
```diff
@@ -3,13 +3,52 @@ import fetch from "node-fetch";
...
+        }
+        dates = [...dates, new Date(endDate).toISOString().split("T")[0]];
+        return dates;
+      };
+      const start = new Date().toISOString().split("T")[0];
+      const end = new Date(new Date().getTime() + 4 * 24 * 60 * 60 * 1000)
+        .toISOString()
+        .split("T")[0];
+
+      for (let i = 0; i < getDatesBetweenDates(start, end).length; i++) {
```
Comment (richardolsson) [inline src/helpers/screenings.js:46]: When you call a function like this inside the _condition_ of a for loop, it will be evaluated (and hence executed) every time the loop runs. So since it always returns five dates, it will be executed five times here.

It's also executed again on the next line, so that means it's executed 10 times.

That's for every screening, so if you retrieve 25 screenings, that's 10 * 25 times in total that `getDatesBetweenDates()` will be called.

Inside `getDatesBetweenDates()` there is another loop, that is 4 iterations long (if I'm not mistaken). That means that in total, the code on lines 35-36 is executed 1000 times (10 * 25 * 4) whenver `screeningsStartpage()` is called.

That's not a very efficient algorithm. But it works! 😊

File: src/helpers/screenings.js [hunk 4]
```diff
@@ -3,13 +3,52 @@ import fetch from "node-fetch";
...
+        if (screenings.start_time.includes(getDatesBetweenDates(start, end)[i])) {
```
Comment (richardolsson) [inline src/helpers/screenings.js:47]: There are easier ways of doing this. I would use a completely different logic, using `Date.getTime()` and simple comparison with the `<` and `>` operators.

But if you do want to use this approach of creating an array of dates, and seeing if a certain date is among those dates, an easier way would be to just do `getDatesBetweenDates().includes(screenings.start_time)`, without the for loop.

You would of course then have to make sure that the format is identical for all of the data in the array, and `screenings.start_time`.

File: src/helpers/screenings.js [hunk 5]
```diff
@@ -3,13 +3,52 @@ import fetch from "node-fetch";
...
+  const result = payload.data
+    .map((item) => ({
+      id: item.id,
+      ...item.attributes,
+      movie: {
+        id: item.attributes.movie.data.id,
+        ...item.attributes.movie.data.attributes,
+      },
+    }))
+    .filter((screenings) => {
```
Comment (richardolsson) [inline src/helpers/screenings.js:29]: This variable (`screenings`) will get the value of a single screening, and be executed once for every screening in the `payload.data` array. So I think `screening` (singular) would be a better name.

File: src/routes/pages.routes.js [hunk 6]
```diff
@@ -1,139 +1,111 @@
-import { Router } from 'express';
-import { loadMovies, loadMovie } from '../helpers/movies.js';
-import { marked } from 'marked';
-import menu from '../../menu.js';
+import { Router } from "express";
```
Comment (richardolsson) [inline src/routes/pages.routes.js:1]: There are a lot of changes in this file and others, that seems to be mostly unrelated to the work that you've done. Avoid contaminating your PR with unrelated changes by making sure everyone in the team uses auto formatting withthe same editor settings.

File: static/js/movies.js [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+    }
+
+    const li = document.createElement("li");
+    li.innerHTML = `<p class="date">${date}</p>
+    <a href="/movies/${movie.id}">
+    <p class="time">${time}</p>
+    <img src="${movie.image.url}"/>
+    ${movie.title.slice(0, 17)}</a>`;
+
+    switch (true) {
```
Comment (richardolsson) [inline static/js/movies.js:36]: I'm not a fan of this construct at all. Simple if statements are usually the best, IMHO.

File: static/js/movies.js [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+      case date.includes(new Date().toISOString().split("T")[0]):
```
Comment (richardolsson) [inline static/js/movies.js:37]: As mentioned elsewhere, it would probably be easier to do date comparisons using `Date.getTime()` and regular `<` or `>`.

**Review by richardolsson — COMMENTED** [review]
This was a big task, that I think you solved well. I've left a bunch of feedback below but no changes are needed.
